src/Psl/Type/Internal/DictType.php

```diff
@@ -39,21 +39,31 @@ public function __construct(
     public function coerce(mixed $value): array
     {
         if (! is_iterable($value)) {
-            throw CoercionException::withValue($value, $this->toString(), $this->getTrace());
+            throw CoercionException::withValue($value, $this->toString());
         }
 
-        $trace = $this->getTrace();
-        $key_type = $this->key_type->withTrace($trace->withFrame('dict<' . $this->key_type->toString() . ', _>'));
-        $value_type = $this->value_type->withTrace($trace->withFrame('dict<_, ' . $this->value_type->toString() . '>'));
-
         $result = [];
+        $key_type = $this->key_type;
+        $value_type = $this->value_type;
 
         /**
          * @var Tk $k
          * @var Tv $v
          */
         foreach ($value as $k => $v) {
-            $result[$key_type->coerce($k)] = $value_type->coerce($v);
+            try {
+                $k_result = $key_type->coerce($k);
+            } catch (CoercionException $e) {
+                throw CoercionException::withValue($k, $this->toString(), 'key(' . (string) $k . ')', $e);
+            }
+
+            try {
+                $v_result = $value_type->coerce($v);
+            } catch (CoercionException $e) {
+                throw CoercionException::withValue($v, $this->toString(), (string) $k, $e);
+            }
+
+            $result[$k_result] = $v_result;
         }
```

Comment:
just a suggestion, wouldn't this be more performant? haven't tested it, but i think a having a single try-catch would be better here ( same for other types were we iterate over values ).

```php
$k = $v = null;
$tk = true; // "tk = trying key"
try {
  foreach ($value as $k => $v) {
    $tk = true;
    $k_result = $key_type->coerce($k);
    $tk = false;
    $v_result = $key_type->coerce($v);
            
    $result[$k_result] = $v_result;
  }
} catch (CoercionException $e) {
  $ts = $this->toString(); // "ts = type string"
  if ($tk) {
    throw CoercionException::withValue($k, $ts, 'key(' . (string) $k . ')', $e);
  } else {
    throw CoercionException::withValue($v, $ts, (string) $k, $e);
  }
}
```

Reply:
much better indeed! (see new included benchmarks)